components/binance/browser/binance_service.cc

```diff
@@ -0,0 +1,486 @@
+
+#include "brave/components/binance/browser/binance_service.h"
+
+#include <algorithm>
+#include <string>
+#include <utility>
+
+#include "base/base64.h"
+#include "base/bind.h"
+#include "base/containers/flat_set.h"
+#include "base/files/file_enumerator.h"
+#include "base/files/file_util.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "base/time/time.h"
+#include "base/token.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/binance/browser/binance_json_parser.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/country_codes/country_codes.h"
+#include "components/os_crypt/os_crypt.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+#include "crypto/random.h"
+#include "crypto/sha2.h"
+#include "net/base/load_flags.h"
+#include "net/base/url_util.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+
+namespace {
+
+const char client_id[] = BINANCE_CLIENT_ID;
+const char oauth_host[] = "accounts.binance.com";
+const char api_host[] = "api.binance.com";
+const char oauth_callback[] = "com.brave.binance://authorization";
+const char oauth_scope[] =
+    "user:email,user:address,asset:balance,asset:ocbs";
+const GURL oauth_url("https://accounts.binance.com/en/oauth/authorize");
+const unsigned int kRetriesCountOnNetworkChange = 1;
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("binance_service", R"(
+      semantics {
+        sender: "Binance Service"
+        description:
+          "This service is used to communicate with Binance "
+          "on behalf of the user interacting with the Binance widget."
+        trigger:
+          "Triggered by user connecting the Binance widget."
+        data:
+          "Account balance for the widget."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on the new tab page."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+GURL GetURLWithPath(const std::string& host, const std::string& path) {
+  return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+}
+
+std::string GetHexEncodedCryptoRandomSeed() {
+  const size_t kSeedByteLength = 28;
+  // crypto::RandBytes is fail safe.
+  uint8_t random_seed_bytes[kSeedByteLength];
+  crypto::RandBytes(random_seed_bytes, kSeedByteLength);
+  return base::HexEncode(
+      reinterpret_cast<char*>(random_seed_bytes), kSeedByteLength);
+}
+
+std::string GetCodeChallenge(const std::string& code_verifier) {
+  std::string code_challenge;
+  char raw[crypto::kSHA256Length] = {0};
```

Comment:
Just checking: This doesn't need to be `uint8_t`?

Reply:
I think it's ok with char as it avoids a cast later on